bilby/core/prior/dict.py

```diff
@@ -465,39 +466,122 @@ def sample_subset_constrained(self, keys=iter([]), size=None):
+            the integration estimate.
+
+        """
+        keys = tuple(keys)
         if keys in self._cached_normalizations.keys():
             return self._cached_normalizations[keys]
-        else:
-            factor_estimates = [
-                self._estimate_normalization(keys, min_accept, sampling_chunk)
-                for _ in range(nrepeats)
-            ]
-            factor = np.mean(factor_estimates)
-            if np.std(factor_estimates) > 0:
-                decimals = int(-np.floor(np.log10(3 * np.std(factor_estimates))))
-                factor_rounded = np.round(factor, decimals)
-            else:
-                factor_rounded = factor
-            self._cached_normalizations[keys] = factor_rounded
-            return factor_rounded
 
-    def _estimate_normalization(self, keys, min_accept, sampling_chunk):
-        samples = self.sample_subset(keys=keys, size=sampling_chunk)
-        keep = np.atleast_1d(self.evaluate_constraints(samples))
-        if len(keep) == 1:
-            self._cached_normalizations[keys] = 1
-            return 1
-        all_samples = {key: np.array([]) for key in keys}
-        while np.count_nonzero(keep) < min_accept:
-            samples = self.sample_subset(keys=keys, size=sampling_chunk)
-            for key in samples:
-                all_samples[key] = np.hstack([all_samples[key], samples[key].flatten()])
-            keep = np.array(self.evaluate_constraints(all_samples), dtype=bool)
-        factor = len(keep) / np.count_nonzero(keep)
-        return factor
+        sample_keys = [key for key in keys]
+
+        # check if the constraint can be applied to the selection of keys
+        try:
+            sample = self.sample_subset(keys, size=1)
+            self.conversion_function(sample)
+        except KeyError:
+            raise ValueError("'keys' does not contain all parameters needed to evaluate the constraint.")
+
+        for key in sample_keys:
+            if isinstance(self[key], JointPrior):
+                dist_keys = set(self[key].dist.names)
+                missing_keys = list(dist_keys - set(keys))
+                sample_keys.extend(missing_keys)
+
+        if method == "qmc_quad":
+            integrator = self._integrate_normalization_factor_from_qmc_quad
+        elif method == "from_samples":
+            integrator = self._integrate_normalization_factor_from_samples
+        else:
+            raise ValueError(f"Integration method {method} not understood")
+        try:
+            theta = np.random.uniform(0, 1, size=(len(keys), 1))
+            samples = {key: self[key].rescale(units) for key, units in zip(keys, theta)}
+            for key in keys:
+                if np.any(np.isnan(samples[key])):
+                    print("The rescale method appears to be not working. Switching to 'sample_based'.")
+                    integrator = self._integrate_normalization_factor_from_samples
+        except Exception:
```

Comment:
It's probably better completely removed. I thought about user-side errors in the implementation of custom priors, but such cases should probably fail loudly rather than silently.

Maybe catching `NotImplementedError` could be added to prepare for later changes that add `NotImplementedError`s to the base classes, but that would also render the check if some priors yield `None` unnecessary.

Reply:
I updated the PR so that both cases are accounted for